doc/AUTOMATIC_BOOT_ASSESSMENT.md

```diff
@@ -0,0 +1,203 @@
+
+3. The system is booted for the first time after the new kernel is
+   installed. The boot loader now sees the `+3` counter in the entry file
+   name. It hence renames the file to `4.14.11-300.fc27.x86_64+2-1.conf`
+   indicating that at this point one attempt has started and thus only one less
+   is left. After the rename completed the entry is booted as usual.
+
+4. Let's say this attempt to boot fails. On the following boot the boot loader
+   will hence see the `+2-1` tag in the name, and hence rename the entry file to
+   `4.14.11-300.fc27.x86_64+1-2.conf`, and boot it.
+
+5. Let's say the boot fails again. On the subsequent boot the loader hence will
+   see the `+1-2` tag, and rename the file to
+   `4.14.11-300.fc27.x86_64+0-3.conf` and boot it.
+
+6. If this boot also fails, on the next boot the boot loader will see the the
+   tag `+0-3`, i.e. the counter reached zero. At this point the entry will be
+   considered "bad", and ordered to the end of the list of entries. The next
+   newest boot entry is now tried, i.e. the system automatically reverted back
+   to an earlier version.
+
+The above describes the walkthrough when the selected boot entry continously
+fails. Let's have a look at an alternative ending to this walkthrough. In this
+scenario the first 4 steps are the same as above:
+
+1. *as above*
+
+2. *as above*
+
+3. *as above*
+
+4. *as above*
+
+5. Let's say the second boot succeeds. The kernel initializes properly, systemd
+   is started and invokes all generators.
+
+6. One of the generators started is `systemd-boot-complete-generator` which
+   detects that boot counting is used. It hence pulls
+   `systemd-boot-complete.service` into the initial transaction.
+
+7. `systemd-boot-complete.service` is ordered after and `Requires=` the generic
+   `boot-complete.target` unit. This unit is hence also pulled into the initial
+   transaction.
+
+8. The `boot-complete.target` unit is ordered after and pulls in various units
+   that are required to succeed for the boot process to be considered
+   successful. One such unit is `systemd-boot-check.service`.
+
+9. `systemd-boot-check.service` is run after all its own dependencies
+   completed, and assesses that the boot completed successfully. It hence exits
+   cleanly.
+
+10. This allows `boot-complete.target` to be reached. This signifies to the
+    system that this boot attempt shall be considered successful.
+
+11. Which in turn permits `systemd-boot-complete.service` to run. It now
+    determines which boot loader entry file was used to boot the system, and
+    renames it dropping the counter tag. Thus
+    `4.14.11-300.fc27.x86_64+1-2.conf` is renamed to
+    `4.14.11-300.fc27.x86_64.conf`. From this moment boot counting is turned
+    off.
+
+12. On the following boot (and all subsequent boots after that) the entry is
+    now seen with boot counting turned off, no further renaming takes place.
```

Comment:
Maybe consider reseting the counter instead of disabling boot-counting entirely.

I understand the idea of "if it booted once properly, the image is correct and it won't magically stop working" but I have actually seen that happen in real life. 

Suppose you have two completely separate (i.e different rootfs) boot entries. One is the normal operation one, the other is a recovery system that can do only one thing : reformat and reinstall the normal partition.

The normal partition might become bad for various reasons on a completely autonomous system. Bad blocks, bad handling of disk full, bad recovery on incorrect shutdown... Those kind of things are very rare, but on autonomous systems without any sysadmin they can happen and, statistically, they do.

At this point, having the boot counter still active provides a simple "parachute" for a complete reinstall, and that's very welcome.

Reply:
@medhefgo also suggested this, see above. and also see my comments on that. Besides the downgrade-attack thing I also think there's value in putting boundaries on the time where we keep changing the $BOOT file system. With the implemented approach, if an update cycle is clean we'll update $BOOT three times: once to install the new kernel, once on the next boot to count down, and once after it completed to mark it OK. After that $BOOT isn't touched anymore, regardless how many cycles follow, until the next update. I think that's much better behaviour than continuing to count forerver.

I mean, the current scheme makes its changes to the ESP, which is VFAT, and hence has less than ideal integrity behaviour. If we had something more reliable we'd use it, but it's what we have to work with, but we should still optimize the write cycles, and I think O(1) instead of O(n) write cycles per update is highly preferable.

But dunno. we can certainly look into making other schemes workable if there's really the need, once the initial support is in.

Reply:
That's no big deal, I'll have to reimplement that part of the logic for u-boot anyway...

that was more a case of "i've been there before" but there are many use cases and only one can be the default...